### PR TITLE
support lookup on tagName/edgeName

### PR DESCRIPTION
**Pull request by panda-sheep**
The following syntax is supported:
```
lookup on tagName
lookup on edgeName
```
Where and yield clauses are not allowed at this time.

### CODE REVIEW

File: src/optimizer/rule/IndexScanRule.cpp [hunk 1]
```diff
@@ -402,6 +430,16 @@ IndexItem IndexScanRule::findOptimalIndex(graph::QueryContext *qctx,
...
+// Only use "lookup on tagname"
+IndexItem IndexScanRule::findOptimalIndex(graph::QueryContext *qctx,
+                                          const OptGroupNode *groupNode) const {
+    auto indexes = allIndexesBySchema(qctx, groupNode);
+    if (indexes.empty()) {
+        return nullptr;
+    }
+    return indexes[0];
+}
+
```
Comment (bright-starry-sky) [inline src/optimizer/rule/IndexScanRule.cpp:442]: A suggestion . rename the method to `findLightestIndex`, and find the lightest index. for example : 
index1 on tag ()
index2 on tag(c1)
index3 on tag(c1,c2,...)
Index1 should be selected. right?

**Comment by bright-starry-sky** [comment]
Except one suggestion of above, totally is good for me.

**Comment by bright-starry-sky** [comment]
Emm, I think you need to check the yield clause. the yield clause should be empty when the filter is empty. right?

File: src/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (panda-sheep) [reply src/optimizer/rule/IndexScanRule.cpp:442]: The current strategy is to find an index randomly.

**Comment by panda-sheep** [comment]
> Emm, I think you need to check the yield clause. the yield clause should be empty when the filter is empty. right?

Good question， IndexScanValidator::prepareYield has already handled.

**Comment by bright-starry-sky** [comment]
> > Emm, I think you need to check the yield clause. the yield clause should be empty when the filter is empty. right?
> 
> Good question， IndexScanValidator::prepareYield has already handled.

Got. I saw. my mistake.

File: src/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (panda-sheep) [reply src/optimizer/rule/IndexScanRule.cpp:442]: Here, the first index is used as the index to be used for returning.

Comment (bright-starry-sky) [reply src/optimizer/rule/IndexScanRule.cpp:442]: > Here, the first index is used as the index to be used for returning.

I think, the best scanning performance of the storage layer is the lightest index. WDYT?
In addition, if return the first index, this method name is best `getFirstIndex` or `findAnyIndex` . right?

Comment (panda-sheep) [reply src/optimizer/rule/IndexScanRule.cpp:442]: Make sense,  Let’s use the index with the fewest fields, 👍

Comment (panda-sheep) [reply src/optimizer/rule/IndexScanRule.cpp]: Fixed.

**Review by dutor — COMMENTED** [review]
Thanks!

I have some questions about the design:
 1. If the filter is not present, surely we could scan by label. But what if the filter is provided while no applicable index can be used?
 2. Why can't we return properties while scanning by label?

**Review by bright-starry-sky — APPROVED** [review]
LGTM.

**Comment by Shylock-Hg** [comment]
How to scan label with properties filter?

**Comment by panda-sheep** [comment]
> Thanks!
> 
> I have some questions about the design:
> 
> 1. If the filter is not present, surely we could scan by label. But what if the filter is provided while no applicable index can be used?
> 2. Why can't we return properties while scanning by label?


`lookup on tagName  `      Use index to return tag's  all  `vid `by default
lookup on edgeName       Use index to return the `src, rank, dst `   of the all  edges of one edgetype  by default

**Comment by panda-sheep** [comment]
> > Thanks!
> > I have some questions about the design:
> > 
> > 1. If the filter is not present, surely we could scan by label. But what if the filter is provided while no applicable index can be used?
> > 2. Why can't we return properties while scanning by label?
> 
> `lookup on tagName ` Use index to return tag's all `vid `by default
> lookup on edgeName Use index to return the `src, rank, dst ` of the all edges of one edgetype by default
> 
> So no where clause is needed, No need for yield clause

**Comment by dutor** [comment]
> > Thanks!
> > I have some questions about the design:
> > 
> > 1. If the filter is not present, surely we could scan by label. But what if the filter is provided while no applicable index can be used?
> > 2. Why can't we return properties while scanning by label?
> 
> `lookup on tagName ` Use index to return tag's all `vid `by default
> `lookup on edgeName` Use index to return the `src, rank, dst ` of the all edges of one edgetype by default
> 
> So no` where clause` is needed, No need for `yield clause`

You didn't answer any of my questions.

**Comment by panda-sheep** [comment]
> > > Thanks!
> > > I have some questions about the design:
> > > 
> > > 1. If the filter is not present, surely we could scan by label. But what if the filter is provided while no applicable index can be used?

> > > 2. Why can't we return properties while scanning by label?
> > 
ok
1） Before this pr， if the current `lookup `statement does not have a `where `condition, an error will be reported
If a filter is provided without an applicable index, it will report that there is no valid index.

2)  When there is no valid index, we do not scan the data.  I think you may confuse the where statement with the filter.

**Review by dutor — APPROVED** [review]

**Pull request merged**